File: src/game/components/Camera.ts
import { Mediator } from "../core/events/Mediator";
import { Character } from "./Character";
import { Ground } from "./Ground";
import { Direction } from "../managers/KeyboardManager";
import { AnimationManager } from "../core/animations/AnimationManager";
import { GameWindow } from "../core/renderer/GameWindow";

export class Camera {
    private positionX: number = 0;
    private positionY: number = 0;

    private realX: number = 0;
    private realY: number = 0;

    public maxX: number = 0;
    public maxY: number = 0;

    public animationLock: boolean = false;

    private character: Character | undefined;
    private ground: Ground | undefined;

    constructor(
        private mediator: Mediator,
        private gameWindow: GameWindow,
        private animationManager: AnimationManager
    ) {
        this.mediator.registerHandler('Character::ChangedDirection', async (direction: Direction) => {
            await this.move(direction);
        });
    }

    public async move(direction: Direction) {
        if (direction === null) {
            return;
        }

        if (direction === 'up' || direction === 'left' || direction === 'right' || direction === 'down') {
            return;
        }

        if (this.animationLock) {
            return;
        }

        this.character.move(direction);

        if (direction === 'cam_down') {
            this.positionY -= 2;
        }

        if (direction === 'cam_right') {
            this.positionX -= 2;
        }

        if (direction === 'cam_up') {
            this.positionY += 2;
        }

        if (direction === 'cam_left') {
            this.positionX += 2;
        }

        const realX = this.positionX * 32;
        const realY = this.positionY * 32;

        this.animationLock = true;

        await this.animationManager.animateTo(this, { realX, realY }, 1000.0 / 6.5, () => {});

        this.animationLock = false;
    }

    public attachCharacter(character: Character) {
        this.character = character;
    }

    public async centerToCharacter() {
        let { positionX, positionY } = this.calculateCameraPosition();

        this.positionX = positionX;
        this.positionY = positionY;
        this.realX = positionX * 32;
        this.realY = positionY * 32;
    }

    public async centerToCharacterAnimated() {
        let { positionX, positionY } = this.calculateCameraPosition();

        if (this.animationLock) {
            if (Math.abs(positionX - this.positionX) > 1 || Math.abs(positionY - this.positionY) > 1) {
                this.animationManager.updateAnimation(this, { positionX, positionY }, 1000.0);
            } else {
                this.animationManager.updateAnimation(this, { positionX, positionY }, 2500.0);
            }

            return;
        }

        this.animationLock = true;

        await this.animationManager.animateTo(this, { positionX, positionY }, 2500.0, (percentage) => {
            this.realX = this.positionX * 32;
            this.realY = this.positionY * 32;
        });

        this.animationLock = false;
    }

    private calculateCameraPosition() {
        const { width, height } = this.gameWindow.dimensions();

        let positionX = -this.character.positionX + (width / 2 / 32) - 1;
        let positionY = -this.character.positionY + (height / 2 / 32) - 1;

        if (positionX > 0) {
            positionX = 0;
        }

        if (positionY > 0) {
            positionY = 0;
        }

        if (positionY < this.maxY) {
            positionY = this.maxY;
        }

        if (positionX < this.maxX) {
            positionX = this.maxX;
        }

        return {
            positionX,
            positionY
        };
    }

    public attachGround(ground: Ground) {
        this.ground = ground;

        const { width, height } = this.gameWindow.dimensions();

        const x = (width) / 32;
        const y = (height) / 32;

        let offsetX = 0;
        let offsetY = 0;

        if (width > this.ground.width * 32) {
            offsetX = (width - (this.ground.width * 32)) / 2 / 32;
        }

        if (height > this.ground.height * 32) {
            offsetY = (height - (this.ground.height * 32)) / 2 / 32;
        }

        this.maxX = -this.ground.width + x - offsetX;
        this.maxY = -this.ground.height + y - offsetY;
    }

    public getPosition() {
        return {
            cameraRealX: this.realX,
            cameraRealY: this.realY
        }
    }
}

File: Application/Components/Camera.cs
using MoriganBlazorClient.Application.Animations;
using MoriganBlazorClient.Application.Animations.Interfaces;

namespace MoriganBlazorClient.Application.Components
{
    public class Camera : IAnimatedCanvasPosition
    {
        private readonly AnimationManager _animationManager;

        private long _positionX { get; set; }

        private long _positionY { get; set; }

        private long _width { get; set; }

        private long _height { get; set; }

        private long _maxX { get; set; }

        private long _maxY { get; set; }

        public double CanvasPositionX { get; set; }

        public double CanvasPositionY { get; set; }

        public Camera(AnimationManager animationManager)
        {
            System.Console.WriteLine($"{nameof(Camera)}->ctor");
            
            _animationManager = animationManager;
        }

        public void SetWindowDimensions(long width, long height) 
        {
            _width = width;
            _height = height;
        }

        public void CenterToCharacter(Character character) 
        {
            var positionX = -character.PositionX + (_width / 2 / 32) - 1;
            var positionY = -character.PositionY + (_height / 2 / 32) - 1;

            if (positionX > 0) {
                positionX = 0;
            }

            if (positionY > 0) {
                positionY = 0;
            }

            if (positionY < _maxY) {
                positionY = _maxY;
            }

            if (positionX < _maxX) {
                positionX = _maxX;
            }

            _positionX = positionX;
            _positionY = positionY;

            _animationManager.Animate(this, positionX * 32, positionY * 32, 850);
        }

        public void CalculateMaxValues(Ground ground) 
        {
            var (groundWidth, groundHeight) = ground.GetDimensions();

            var x = (_width) / 32;
            var y = (_height) / 32;

            long offsetX = 0;
            long offsetY = 0;

            if (_width > groundWidth * 32) {
                offsetX = (_width - (groundWidth * 32)) / 2 / 32;
            }

            if (_height > groundHeight * 32) {
                offsetY = (_height - (groundHeight * 32)) / 2 / 32;
            }

            _maxX = -groundWidth + x - offsetX;
            _maxY = -groundHeight + y - offsetY;
        }
    }
}